docs/languages/csharp/async.md

```diff
@@ -2,44 +2,130 @@
-*   When the `await` keyword is applied, it suspends the calling method and yields control back to its caller until the awaited task is complete. This is what allows a UI to be responsive and a service to be elastic.
+There's a lot of moving pieces where asynchronous operations are concerned.  If you're curious about what's happening underneath the covers of `Task` and `Task<T>`, checkout the [Async in-depth](async-in-depth.md) article for more information.
+
+On the C# side of things, the compiler transforms your code into a state machine which keeps track of things like yielding execution when an `await` is reached, resuming execution when a background job has finished, and so on.
+
+For the theoretically-inclined, this is an implementation of the [Promise Model of asynchrony](https://en.wikipedia.org/wiki/Futures_and_promises).
+
+## Key Pieces to Understand
+
+*   Async code can be used for both I/O-bound and CPU-bound code, but differently for each scenario.
+*   Async code uses `Task<T>` and `Task`, which are constructs used to model work being done in the background.
+*   When the `await` keyword is applied, it suspends the calling method and yields control back to its caller until the awaited task is complete.
 *   `await` can only be used inside an async method.
-*   Unless an async method has an `await` inside its body, it will never yield!
-*   `async void` should **only** be used on Event Handlers (where it is required).
 
-## Example (C#)
+## Recognize CPU-Bound and I/O-Bound Work
+
+The first two examples of this guide showed how you can use `async` and `await` for I/O-bound and CPU-bound work.  It's key that you can identify when a job you need to do is I/O-bound or CPU-bound, because it can greatly affect the performance of your code and could potentially lead to misusing certain constructs.
+
+Here are two questions you should ask before you write any code:
+
+1. Will my code be "waiting" for something, such as data from a database?
+
+    If your answer is "yes", then your work is **I/O-bound**.
+
+2. Will my code be performing an expensive computation?
+
+    If you answered "yes", then your work is **CPU-bound**.
+    
+If the work you have is **I/O-bound**, use `async` and `await` *without* `Task.Run`.  You *should not* use the Task Parallel Library.  The reason for this is outlined in the [Async in Depth article](../../async/async-in-depth.md).
 
-The following example shows how to write basic async code for both a client app and a web service. The code, in both cases, will count the number of times ”.NET” appears in the HTML of “dotnetfoundation.org”.
+If the work you have is **CPU-bound** and you care about responsiveness, use `async` and `await` but spawn the work off on another thread *with* `Task.Run`.  If the work can be parallelized, you should also consider using the Task Parallel Library.
```

Comment:
This might be misleading. You can get into situations (like a web server that does CPU bound work) where the overhead of thread switching overwhelms any benefit of doing the work on another thread...